src/coreComponents/physicsSolvers/multiphysics/CoupledReservoirAndWellsBase.cpp

```diff
@@ -110,6 +110,47 @@ addCouplingNumNonzeros( SolverBase const * const solver,
   } );
 }
 
+bool validateWellPerforations( SolverBase const * const solver,
+                               WellSolverBase const * const wellSolver,
+                               DomainPartition const & domain )
+{
+  std::pair< string, string > badPerforation;
+  solver->forDiscretizationOnMeshTargets( domain.getMeshBodies(), [&] ( string const &,
+                                                                        MeshLevel const & meshLevel,
+                                                                        arrayView1d< string const > const & regionNames )
+  {
+    ElementRegionManager const & elemManager = meshLevel.getElemManager();
+    elemManager.forElementSubRegions< WellElementSubRegion >( regionNames, [&]( localIndex const, WellElementSubRegion const & subRegion )
+    {
+      PerforationData const * const perforationData = subRegion.getPerforationData();
+      WellControls const & wellControls = wellSolver->getWellControls( subRegion );
+
+      arrayView1d< localIndex const > const & resElementRegion =
+        perforationData->getField< fields::perforation::reservoirElementRegion >();
+
+      // Loop over perforations and check the reservoir region to which each perforation is connected to
+      // If the name of the region is not in the list of targetted regions, then we have a "bad" connection.
+      for( localIndex iperf = 0; iperf < perforationData->size(); ++iperf )
+      {
+        localIndex const er = resElementRegion[iperf];
+        string const regionName = elemManager.getRegion( er ).getName();
+        if( std::find( regionNames.begin(), regionNames.end(), regionName ) == regionNames.end())
+        {
+          badPerforation = {wellControls.getName(), regionName};
+          break;
+        }
```

Comment:
and then, if you decide to follow the first suggestion, you can just search for the `regionName` in the target regions of the flow solver:
```suggestion
        if( std::find( flowTargetRegionNames.begin(), flowTargetRegionNames.end(), regionName ) == flowTargetRegionNames.end())
        {
          badPerforation = {wellControls.getName(), regionName};
          break;
        }
```
where `flowTargetRegionNames` would have been previously defined as:
```
arrayView1d< string const > const flowTargetRegionNames = solver->getReference< string >( viewKeyStruct::targetRegionsString );
```
With this approach, you would save a little bit by searching for `regionName` in a smaller array that does not contain the well regions. This would require to slightly change the arguments you pass to the function (`solver` is no longer the coupled solver).

But I have not tried, so there may be something that I am missing.

Reply:
Make sense - and works. I have reordered the loop using the reservoir solver regions as the lookup list.